yaml.qmd

```diff
@@ -0,0 +1,874 @@
+
+A similar problem occurs when you nest options incorrectly.
+For example, you might get `code-fold` right, but forget to nest `html` under `format`:
+
+``` {.yaml filename="document.qmd"}
+---
+title: My document
+html:
+  code-fold: true
+---
+```
+
+As a top-level option `html` isn't something Quarto knows about, but it passes validation.
+The result is no error, and no folded code.
+
+[Code completion](#yaml-completion) can be a useful indicator for these problems.
+If you aren't getting the completion you expect, perhaps you've got the wrong name, or you've nested it under the wrong option.
+
+### Specifying the wrong type of value
+
+Checking the type of value you've passed is something Quarto is good at, assuming your syntax is correct and you've got the right option name.
+
+For example, consider trying to add a table of contents like this:
+
+``` {.yaml .error filename="document.qmd"}
+---
+title: My document
+toc: yes
+---
+```
+
+On render Quarto returns the error:
+
+``` default
+ERROR: Validation of YAML front matter failed.
+ERROR: (line 3, columns 6--8) Field "toc" has value yes, which must instead be `true` or `false`
+
+✖ The value yes is a string.
+ℹ The error happened in location toc.
+ℹ Quarto uses YAML 1.2, which interprets booleans strictly.
+ℹ Try using true instead.
+
+ERROR: Render failed due to invalid YAML
+```
+
+This is a really useful message---it tells you the YAML front matter validation failed; it tells you where the problem is, `(line 3, columns 6--8) Field "toc"`; and it tells you what the problem was, Quarto expected a boolean, but got a string.
+It even suggests a solution `try using true instead`.
+
+Not all validation error messages are as illuminating.
+For example, consider trying to add an "Other Link" like this:
+
+``` {.yaml .error filename="document.qmd"}
+---
+title: "Untitled"
+format:
+  html:
+    other-links: http://quarto.org
+---
+```
+
+The validation error returned is:
+
+``` default
+ERROR: Validation of YAML front matter failed.
+ERROR: In file type-errors.qmd
+(line 5, columns 18--34) Field "other-links" has value http://quarto.org, which must instead be 'false'
+4:   html:
+5:     other-links: http://quarto.org
+                   ~~~~~~~~~~~~~~~~~
+6: ---
+```
+
+It's clear where the error is (`(line 5, columns 18--34) Field "other-links"`), and that the wrong type of value has been encountered.
+But it's unclear what Quarto actually wants.
+It says it was expecting `false`, but that doesn't help us figure out how to actually specify a link.
+
+::: callout-tip
+## Obtuse messages
+
+Quarto tries not to overwhelm you with output in an error message, but sometimes the process of shortening an error message leads to a message that isn't very helpful.
+
+This particular form of message that ends with `which must instead be x`, is a common culprit of being too brief to be helpful.
+Quarto will often report one value `x`, but in reality there are many values that the option could take.
```

Comment:
Is the suggested `x` usually the first alphabetically among the options? Is that worth noting here?